Source/DafnyLanguageServer/Workspace/ChangeProcessors/TextChangeProcessor.cs

```diff
@@ -5,29 +5,38 @@
 
 namespace Microsoft.Dafny.LanguageServer.Workspace.ChangeProcessors {
   public class TextChangeProcessor : ITextChangeProcessor {
-    public TextDocumentItem ApplyChange(TextDocumentItem originalTextDocument, DidChangeTextDocumentParams documentChange, CancellationToken cancellationToken) {
+    public DocumentTextBuffer ApplyChange(DocumentTextBuffer originalTextDocument, DidChangeTextDocumentParams documentChange, CancellationToken cancellationToken) {
+      var newText = ApplyTextChanges(originalTextDocument.Text, originalTextDocument.NumberOfLines, documentChange.ContentChanges, out var newNumberOfLines, cancellationToken);
       return originalTextDocument with {
         Version = documentChange.TextDocument.Version,
-        Text = ApplyTextChanges(originalTextDocument.Text, documentChange.ContentChanges, cancellationToken)
+        Text = newText,
+        NumberOfLines = newNumberOfLines
       };
     }
 
-    private static string ApplyTextChanges(string originalText, Container<TextDocumentContentChangeEvent> changes, CancellationToken cancellationToken) {
+    private static string ApplyTextChanges(string originalText, int originalLines, Container<TextDocumentContentChangeEvent> changes, out int numberOfLines, CancellationToken cancellationToken) {
```

Comment:
I think we can clean up these APIs by letting DafnyDocument not contain a `TextDocumentItem` but instead a `VersionedTextDocumentIdentifier` (containing the Version and Uri) and a `DocumentTextBuffer` (containing just the text and the # of lines), and then these APIs could take and return a `DocumentTextBuffer`.

But we might want to do this in a follow-up PR (or even better, an earlier one), since these changes are all a bit out of scope of hover.

At some point we should also move `PositionsExtensions.ToAbsolutePosition` to `DocumentTextBuffer` and implement it in constant time.